### ISSUE TITLE
[HOLD for payment 2023-08-03] [$1000] You can add newlines in task titles

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23025]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Click + icon and select assign a task
2. Enter title and assign it to a user
3. Click Confirm Task

## Expected Result:
User should not be able to add newlines in task title

## Actual Result:
User is able to add new lines

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** Dev  1.3.41-2
**Reproducible in staging?:** no
**Reproducible in production?:** no
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/17e686df-3f5d-4096-bb2a-c8495c481390

**Expensify/Expensify Issue URL:**
**Issue reported by:** @gadhiyamanan
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1689431653449869

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019beadd628d4e1003</li>
        <li>Upwork Job ID: 1681984801194803200</li>
        2023-07-20<li>Automatic offers: </li>
<ul>
<li> |  | 0</li>
<li>dhairyasenjaliya | Contributor | 25728028</li>
<li>gadhiyamanan | Reporter | 25728030</li>
</ul></li>
    </ul>
</details>

**Comment by thienlnam** [comment]
We believe this is a regression of a recent PR since it's only a problem on dev but it's impacting all multi-line inputs for tasks at least

**Comment by thienlnam** [comment]
Yeah I agree, I don't think think task titles should be multi-line

**Comment by Nikhil-Vats** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Multiline title works on Assign task page but not when editing from Confirm task page.

### What is the root cause of that problem?
In [this PR](https://github.com/Expensify/App/pull/22586) we added the props to allow multiline input in some pages but left some other pages.

https://github.com/Expensify/App/blob/0fcfe75f47c8f7e413f1b08885be124d33f9e17d/src/pages/tasks/NewTaskDetailsPage.js#L102-L104

https://github.com/Expensify/App/blob/0fcfe75f47c8f7e413f1b08885be124d33f9e17d/src/pages/tasks/TaskTitlePage.js#L83-L85

There is also an inconsistency here where if title and description are long then we show description in two lines but title in 1 -
https://github.com/Expensify/App/blob/0fcfe75f47c8f7e413f1b08885be124d33f9e17d/src/pages/tasks/NewTaskPage.js#L158-L171

<details>
<summary>
See screenshot -
</summary>

<img width="375" alt="image" src="https://github.com/Expensify/App/assets/31413407/a3fa87d4-2a29-4ff3-b4b8-836ce204ea6e">

</details>

### What changes do you think we should make in order to solve the problem?
1. We need to remove the `autoGrowHeight`, `textAlignVertical` and `containerStyles` prop from `NewTaskDetailsPage` and `TaskTitlePage`.

2. To fix the other inconsistency where title is shown in 1 line but description in 2, we need to add following props for `MenuItemWithTopDescription` in `NewTaskPage` that shows `title` after line 162 -
    ```javascript
    numberOfLinesTitle={2}
    titleStyle={styles.flex1}
    ``` 

3. We should change `onEntryTransitionEnd={() => focusAndUpdateMultilineInputRange(inputRef.current)}` to `onEntryTransitionEnd={() => inputRef.current && inputRef.current.focus()}` in `TaskTitlePage` as the input is not multiline anymore.

### What alternative solutions did you explore? (Optional)
NA

**Comment by Santhosh-Sellavel** [comment]
@thienlnam What's the expected OP now?

Siongle line for task title right or Can we have a character limit?

**Comment by Nikhil-Vats** [comment]
@Santhosh-Sellavel, they answered [here](https://github.com/Expensify/App/issues/23025#issuecomment-1638634707).

**Comment by Santhosh-Sellavel** [comment]
Thanks @Nikhil-Vats I've checked that, but will wait for @thienlnam here!

**Comment by thienlnam** [comment]
Sorry, I should have clarified. You shouldn't be able to add new lines to the task title. Multi-line tasks are okay if it's a long task title, but it's not okay to add line breaks

**Comment by Santhosh-Sellavel** [comment]
Thank you @thienlnam

**Comment by Santhosh-Sellavel** [comment]
Expected OP

> You shouldn't be able to add new lines to the task title. Multi-line tasks are okay if it's a long task title

Existing proposal does not address this, So I think this is open to proposals. If I missed any please let me know, thanks!

**Comment by Nikhil-Vats** [comment]
@Santhosh-Sellavel my proposal works as well and fixes other inconsistencies.

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we press enter in the task title, we add a newline instead of focusing on the next input

### What is the root cause of that problem?

The default behaviour of textInput is to add a newline when the user press enter

### What changes do you think we should make in order to solve the problem?

To not add a newline when the user press enter, we need to pass 
`blurOnSubmit = {true} `
to the title TextInput.

But we want the next TextInput to be focused so we can do it like this:
1) Create another ref for the description TextInput
`const inputRefDescription = useRef();`
```
<TextInput
        ref={(el) => (inputRefDescription.current = el)}
```
2) In the  title TextInput handle the onBlur event and focus the description input:
`onBlur={() => inputRefDescription.current.focus()}`



### What alternative solutions did you explore? (Optional)

**Comment by ShogunFire** [comment]
@Santhosh-Sellavel Hello, I made a proposal, actually now that I think about it, maybe the alternative solution is better, tell me what you think

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/23025#issuecomment-1646056028)

**Comment by ShogunFire** [comment]
My proposal also works to prevent adding newline with a copy paste

**Comment by Santhosh-Sellavel** [comment]
Task title can be multiple lines, but we should not allow new line so both of your proposal doesn't work @Nikhil-Vats @dhairyasenjaliya 

@ShogunFire Your [proposal](https://github.com/Expensify/App/issues/23025#issuecomment-1646056028) look like a hacky one. 

Will wait for other proposals

**Comment by ShogunFire** [comment]
@Santhosh-Sellavel I am not sure there is a non hacky way to do this.
I tried with blurOnSubmit but it looks bad and with onKeyPress (I stopped in the middle because that would still allow copy pasting newline)
Here is a link to the issue on react native repo and they don't have a fix: https://github.com/facebook/react-native/issues/28535#issuecomment-611196270

**Comment by thienlnam** [comment]
The behavior in the video looks okay to me - I don't think we need to show the input in multiple lines just because we don't accept new lines in that input

**Comment by ShogunFire** [comment]
It looks like this undo what was purpusefully done in other recent PRs
![image](https://github.com/Expensify/App/assets/19537677/c49757fc-982a-4b49-a59a-48dcf1c3c6b2)

My proposal works fine, I can make it feel less hacky with a regex or even modify the  code and make an upstream PR to see what they think

**Comment by thienlnam** [comment]
> It looks like this would undo what was purpusefully done in other recent PRs

Yeah this is intended, we talked about it here 

https://expensify.slack.com/archives/C049HHMV9SM/p1689781369210339

**Comment by Santhosh-Sellavel** [comment]
> The behavior in the video looks okay to me - I don't think we need to show the input in multiple lines just because we don't accept new lines in that input

Thanks, @thienlnam, if that's both @dhairyasenjaliya [proposal](https://github.com/Expensify/App/issues/23025#issuecomment-1638662196) & @Nikhil-Vat's [proposals](https://github.com/Expensify/App/issues/23025#issuecomment-1642643385) are good here.

IMO We should go with @dhairyasenjaliya as their's is first here

> We can reduce number of lines to display the title as well currently its set to show 3 lines on the Task view

Should we do this?

C+ Reviewed
🎀 👀 🎀

**Comment by Santhosh-Sellavel** [comment]
@thienlnam If you have the bandwidth, can you pick this up as CME as you are more aware of the context, thanks!

**Comment by thienlnam** [comment]
Sure, can do!

**Comment by ShogunFire** [comment]
I am confused, the slack conversation that @thienlnam put here seems to agree that we should have a multiline input without newline

> > It looks like this would undo what was purpusefully done in other recent PRs
> 
> Yeah this is intended, we talked about it here
> 
> https://expensify.slack.com/archives/C049HHMV9SM/p1689781369210339

**Comment by thienlnam** [comment]
> We can reduce number of lines to display the title as well currently its set to show 3 lines on the Task view

The current limit seems fine, for tasks titles that are longer than this we just show the ellipsis correct?

**Comment by ShogunFire** [comment]
I am confused, the slack conversation that you put @thienlnam seems to agree that we should have a multiline input without newline

> > It looks like this would undo what was purpusefully done in other recent PRs
> 
> Yeah this is intended, we talked about it here
> 
> https://expensify.slack.com/archives/C049HHMV9SM/p1689781369210339

**Comment by thienlnam** [comment]
Yeah that's correct - we should be able add a long task title which would then break into multiple lines on the task details page but not be able to add line breaks to the input. That appears to be the solution we're addressing in https://github.com/Expensify/App/issues/23025#issuecomment-1638662196. Do you disagree?

**Comment by ShogunFire** [comment]
Oh... I thought the multiline was about the input when we add a task not the titlle in the  task detail page

**Comment by ShogunFire** [comment]
Basically I was thinking we should keep this instead of the behaviour of the video, I think it looks better: 
![image](https://github.com/Expensify/App/assets/19537677/bc67d01d-17d9-47f5-b33b-c2c3e5e23ff5)

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
>  
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: 
> * [x]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:N/A
> * [x]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
> * [x]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug. N/A
> * [x]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@sonialiap] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

This the [PR](https://github.com/Expensify/App/pull/22586) were this was caused, but it was intentional there, so skipping first there checking.

@thienlnam Do we need Regression tests here, I guess let me know your thoughts thanks?

**Comment by thienlnam** [comment]
I think we're fine to not have regression tests 👍

**Comment by Santhosh-Sellavel** [comment]
All set, thanks!

**Comment by Santhosh-Sellavel** [comment]
@sonialiap Can you summarize payments in a new comment, so I can request on ND, thanks!

**Comment by Santhosh-Sellavel** [comment]
> @sonialiap Can you summarize payments in a new comment, so I can request on ND, thanks!

Bump on this ☝️ 

All this qualifies for a bonus too, thanks!

**Comment by sonialiap** [comment]
So sorry for such a long delay! 😬 

@gadhiyamanan report $250 - paid ✔️ 
@dhairyasenjaliya fix + bonus $1500 - paid ✔️ 
@Santhosh-Sellavel review + bonus $1500 - please request in NewDot

**Comment by Santhosh-Sellavel** [comment]
Requested on ND
